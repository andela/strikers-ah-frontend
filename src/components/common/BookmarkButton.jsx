/* eslint-disable prefer-destructuring */
/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faStar } from '@fortawesome/free-solid-svg-icons';
import '../../styles/css/followBtn.css';
import bookmark from '../../redux/actions/bookmarkArticle';

/**
 * Follow button
 */
export class BookmarkButton extends Component {
  /**
   * @author Jacques Nyilinkindi
   * @param {*} props
   * @returns {*} button
   */
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  /**
   * @author Jacques Nyilinkindi
   * @param {*} slug
   * @returns {*} button
   */
  handleClick() {
    const { bookmark: bookmarkArticle, article } = this.props;
    const { slug } = article;
    bookmarkArticle(slug);
  }

  /**
   * @author Jacques Nyilinkindi
   * @param {*} props
   * @returns {*} button
   */
  render() {
    const activeBtn = this.props.article.bookmark ? 'active' : '';
    return (
      <button
        type="button"
        test-data="bookmarkIcon"
        className="bookmarkIcon"
        onClick={this.handleClick}
      >
        <FontAwesomeIcon className={activeBtn} icon={faStar} />
      </button>
    );
  }
}
export const mapStateToProps = state => ({
  Article: state.Article,
});
export default connect(
  mapStateToProps,
  { bookmark },
)(BookmarkButton);
