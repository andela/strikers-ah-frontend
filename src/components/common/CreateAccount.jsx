import React from 'react';
import { Link } from 'react-router-dom';

const CreateAccount = () => (
  <div className="createAccount">
    Create account <span className="makespace" />
    <Link to="/signup">here</Link>
  </div>
);

export default CreateAccount;
