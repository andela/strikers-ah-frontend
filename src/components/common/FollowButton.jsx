/* eslint-disable prefer-destructuring */
/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import { connect } from 'react-redux';
// import { Redirect } from 'react-router-dom';
import '../../styles/css/followBtn.css';
import {
  followUser,
  unfollowUser,
  checkProfileFollowStatus,
} from '../../redux/actions/followingAction';
/**
 * Follow button
 */
class FollowButton extends Component {
  /**
   * @author frank harerimana
   * @param {*} props
   * @returns {*} button
   */
  constructor(props) {
    super(props);
    this.state = {
      username: 'frank',
    };
    this.handleClick = this.handleClick.bind(this);
  }

  /**
   * @author frank harerimana
   * @param {*} event
   * @returns {*} button
   */
  componentDidMount() {
    const token = localStorage.getItem('token');
    if (token !== null) {
      this.checkStatus();
    }
  }

  /**
   * @author frank harerimana
   * @param {*} event
   * @returns {*} button
   */
  checkUsername() {
    // let username;
    // if (this.props.Article.article.authorid != null) {
    //   username = this.props.articleAuthor.article[0].user.username;
    // } else if (this.props.userProfile != null) {
    // const username = this.props.userProfile.user.profile.username;
    // }
    return this.state.username;
  }

  /**
   * @author frank harerimana
   * @param {*} event
   * @returns {*} button
   */
  checkStatus() {
    const username = this.checkUsername();
    this.props.checkProfileFollowStatus(username);
  }

  /**
   * @author frank harerimana
   * @param {*} event
   * @returns {*} button
   */
  handleClick(event) {
    event.preventDefault();
    if (localStorage.getItem('token') != null) {
      const { status } = this.props.following[0];
      const username = this.checkUsername();
      if (status === 'follow') {
        this.props.followUser(username);
      } else {
        this.props.unfollowUser(username);
      }
    } else {
      // eslint-disable-next-line no-return-assign
      return (window.location.href = '/login');
    }
  }

  /**
   * @author frank harerimana
   * @param {*} props
   * @returns {*} button
   */
  render() {
    const { status } = this.props.following[0];
    return (
      // <div className="follow-btn-div">
      <button
        type="button"
        id={status === 'following' ? 'following-btn' : 'follow-btn'}
        className={this.props.className}
        onClick={this.handleClick}
      >
        {status}
      </button>
      // </div>
    );
  }
}
const mapStateToProps = state => ({
  following: state.following,
  userProfile: state.userProfile,
  Article: state.Article,
});

export default connect(
  mapStateToProps,
  { followUser, unfollowUser, checkProfileFollowStatus },
)(FollowButton);
