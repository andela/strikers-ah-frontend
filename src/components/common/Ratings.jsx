/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import '../../styles/css/emoji.css';
import { checkRatings } from '../../redux/actions/rateArticle';

/**
 * Rating Component
 * @returns {*} c
 */
export class Ratings extends Component {
  /**
   *
   * @param {*} props
   */
  constructor(props) {
    super(props);

    this.props.checkRatings(this.props.slug);
  }

  /**
   * @returns {*} render
   */
  render() {
    let ratings = '';
    const { rateArticle } = this.props;
    if (rateArticle.ratings != null) {
      ratings = `${rateArticle.ratings} Average rating`;
    }
    return (
      <div className="ratings">
        <span id="ratings-btn" className={ratings}>
          {ratings}
        </span>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  rateArticle: state.rateArticle,
});

export default connect(
  mapStateToProps,
  { checkRatings },
)(Ratings);
