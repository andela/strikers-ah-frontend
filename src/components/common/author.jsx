import React from 'react';
import { Link } from 'react-router-dom';

const Author = ({ firstname, lastname, username, image, bio }) => {
  return (
    <Link className="author" to={`/${username}`}>
      <div className="author__image">
        <figure className="author__image--figure">
          <img src={image} alt="" />
        </figure>
      </div>
      <div className="author__info">
        <h3 className="author__name">
          {`${firstname === null ? '' : firstname} ${
            lastname === null ? '' : lastname
          }`}
        </h3>
        <h4 className="author__username">@{username}</h4>
        <p className="author__bio">{bio}</p>
      </div>
    </Link>
  );
};

export default Author;
