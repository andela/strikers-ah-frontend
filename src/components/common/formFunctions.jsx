import { Component } from 'react';
import Joi from 'joi-browser';

/**
 * @author Clet Mwunguzi
 * @returns {*} errors
 */
class FormFunctions extends Component {
  /**
   * @author Clet Mwunguzi
   * @returns {*} data
   */
  validate = () => {
    const { data } = this.props;
    const { error } = Joi.validate(data, this.schema, { abortEarly: false });

    if (!error) return null;
    const errors = {};

    // eslint-disable-next-line no-restricted-syntax
    for (const item of error.details) {
      errors[item.path[0]] = item.message;
      if (
        item.path[0] === 'password' &&
        item.message !== '"password" is required'
      ) {
        errors[item.path[0]] =
          '"Password" should be at 8 charcters minimum , with one capital letter, one special character, and a number';
      }
    }

    return errors;
  };

  // eslint-disable-next-line consistent-return
  handleSubmit = e => {
    const { stateChange } = this.props;
    e.preventDefault();

    const error = this.validate();
    stateChange({ errors: error || {} });
    if (error) return null;
    this.submit();
  };

  handleChange = ({ currentTarget }) => {
    const { errors, stateChange } = this.props;
    const message = this.validateInput(currentTarget);
    if (message) errors[currentTarget.name] = message;
    else delete errors[currentTarget.name];

    const { data } = this.props;
    data[currentTarget.name] = currentTarget.value;
    stateChange({ data, errors });
  };

  /**
   * @author Clet Mwunguzi
   * @returns {*} data or errors
   */
  validateInput({ name, value }) {
    const obj = { [name]: value };
    const schema = { [name]: this.schema[name] };
    const { error } = Joi.validate(obj, schema);

    if (name === 'password' && error === null) {
      return null;
    }

    if (
      name === 'password' &&
      error.details[0].message !== '"password" is not allowed to be empty'
    ) {
      return ' "Password" should be at least 8 characters minimum , with one capital letter, one special character, and a number';
    }

    return error ? error.details[0].message : null;
  }
}

export default FormFunctions;
