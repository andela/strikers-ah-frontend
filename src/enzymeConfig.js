import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-localstorage-mock';

export default Enzyme.configure({
  adapter: new Adapter(),
});
