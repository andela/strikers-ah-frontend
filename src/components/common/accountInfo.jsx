import React from 'react';
import { Link } from 'react-router-dom';
import ProfileButton from './profileButton';
import FollowButton from './FollowButton';
import UploadImageForm from './UploadImageForm';
import '../../styles/css/profile.css';

const AccountInfo = ({
  image,
  name,
  bio,
  totalFollowers,
  totalFollowings,
  username,
  owner,
  handleClick,
  showAddImageForm,
  articles,
  currentValues: currentUserInfo,
}) => {
  return (
    <div className="user-account-info" test-data="accountInfoComponent">
      <div className="user-account-flex">
        <div className="img-container">
          <img src={image} alt="" />
          {owner && (
            <button
              className="add-img-toggler"
              type="button"
              onClick={showAddImageForm.toggleForm}
            >
              +
            </button>
          )}
        </div>

        <div>
          <div className="info-container">
            <span className="name">{name.trim().length ? name : username}</span>
            <br />
            <span className="username">{`@${username ||
              name.split(' ')[0]}`}</span>
            {owner ? (
              <ProfileButton label="Edit Profile" handleClick={handleClick} />
            ) : (
              <FollowButton className="btn btn-follow" />
            )}
            <br />
            <span className="bio">{bio}</span>
          </div>
        </div>
      </div>
      <div className="statistics">
        <Link to="/profile" className="stat-label">
          <span className="label">Followers</span>
          <br />
          <span className="value">{totalFollowers}</span>
        </Link>
        <Link to="/profile" className="stat-label">
          <span className="label">Following</span>
          <br />
          <span className="value">{totalFollowings}</span>
        </Link>
        <Link to="/profile" className="stat-label">
          <span className="label">Articles</span>
          <br />
          <span className="value">{articles.length}</span>
        </Link>
      </div>
      {owner && showAddImageForm.visible && (
        <UploadImageForm
          toggleForm={showAddImageForm.toggleForm}
          currentUserInfo={currentUserInfo}
        />
      )}
    </div>
  );
};

export default AccountInfo;
