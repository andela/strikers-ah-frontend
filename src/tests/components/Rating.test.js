import React from 'react';
import { mount } from 'enzyme';
import '../../enzymeConfig';
import { Ratings, mapStateToProps } from '../../components/common/Ratings';

let props = {
  checkRatings: jest.fn(),
  rateArticle: jest.fn(),
  componentWillUpdate: jest.fn(),
  slug: 'this-is-props-slug',
};

let wrapper = mount(<Ratings {...props} />);

describe('Rate.jsx', () => {
  test('should be rendered', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should be able to call the checkRatings', () => {
    props = {
      checkRatings: jest.fn(),
      rateArticle: {
        ratings: '',
      },
      slug: 'this-is-props-slug',
    };
    wrapper = mount(<Ratings {...props} />);
    expect(props.checkRatings).toHaveBeenCalled();
  });

  test('should call the function ', () => {
    const state = {
      rateArticle: {
        rate: {
          error: '',
        },
      },
    };
    mapStateToProps(state);
  });
});
