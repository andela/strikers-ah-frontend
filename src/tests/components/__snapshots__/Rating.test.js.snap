// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rate.jsx should be rendered 1`] = `ReactWrapper {}`;
